src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs

```diff
@@ -73,8 +82,20 @@ public static Task SerializeAsync(Stream utf8Json, object? value, Type inputType
             return WriteAsyncCore<object>(utf8Json, value!, inputType, options, cancellationToken);
         }
 
-        private static async Task WriteAsyncCore<TValue>(Stream utf8Json, TValue value, Type inputType, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        private static async Task WriteAsyncCore<TValue>(
+            Stream utf8Json,
+            TValue value,
+            Type inputType,
+            JsonSerializerOptions? options,
+            CancellationToken cancellationToken)
         {
+            // We flush the Stream when the buffer is >=90% of capacity.
+            // With the default buffer size of 16K, if a single JSON value is >=1.6K it is possible for
+            // the buffer to max out before it can be flushed, causing the buffer to expand.
```

Comment:
I'm still trying to understand the magic behind the 90% number.  The sentence would make just as much sense if it said "We flush the Stream when the buffer is >= 80% of capacity ... if a single JSON value is >= 3.2K it is possible..."... so how was the 90% chosen?  Below it says it's a compromise... was this just based on experimentation and this value found to be ideal for a wide-variety of scenarios?

Reply:
I'll look at adding some more explanation.

Note that we do check for flushing as much as reasonable since we check for flushing after every property and element is written to the buffer, not whole objects or whole arrays.

Background on tradeoffs:
- The initial buffer size (smaller takes up less RAM). We are currently optimizing for payloads < 16K (the default buffer size).
- Minimizing buffer doubling. There are really two reasons why the buffer is doubled:
  - To get the buffer big enough to handle the largest single element\property. We haven't received any direct reports of this threshold causing perf issues.
  - To deal with the buffer being partially used from prior elements\properties. The current algorithm is flush at >= .9.
- Buffer utilization % -- i.e. flushing on every small property write would be wasteful for memory utilization plus the underlying Stream may not appreciate having Flush() called so often.

Consider a scenario of reading in large sets of fixed-size elements, such as bytes arrays. Assuming a single JSON element fits into the current buffer already:
- If threshold is 1-(1\(2^0)) == 0 there would be no doubling (flush every time)
- If threshold is 1-(1\(2^1)) == .5 there would at most 1 doubling
- If threshold is 1-(1\(2^2)) == .75 there would at most 2 doublings
- If threshold is 1-(1\(2^3)) == .875 there would at most 3 doublings
- If threshold is 1-(1\(2^4)) == .9375 there would at most 4 doublings

So at .9 we sit in the "4 doublings" which also leaves some wiggle room for JSON token overhead and other smaller, "normal" properties\elements.

Some other options considered at the time the original todo was created:
- Add a flush threshold knob (in addition to the existing buffer size knob) on the options class.
- Add a heuristic and auto-control the % based on that (such as guessing based on largest prior size).

Both seem like overkill because:
- There is already a knob on the options class to specify the default buffer size. It is useful to increase this if you know certain payloads are going to be large to avoid doubling or to adjust to the underlying Stream's optimal threshold or to match it's internal buffering.
- Since the buffer size is automatically doubled, the buffer will ultimately expand a max of 4 times (for fixed-size elements\properties) which seems like a reasonable number.

Finally we also considered a callback mechanism to have the caller do the flush. So far we haven't received any requests for that (AFAIK).

Reply:
Thanks for the details.